entry/src/block_component.rs

```diff
@@ -0,0 +1,2149 @@
+
+    /// Serializes to bytes with version prefix.
+    fn to_bytes(&self) -> Result<Vec<u8>, bincode::Error> {
+        let marker_bytes = match self {
+            Self::V1(marker) | Self::Current(marker) => marker.to_bytes(),
+            Self::V2(marker) => marker.to_bytes(),
+        }?;
+
+        let mut buffer = Vec::with_capacity(2 + marker_bytes.len());
+        buffer.extend_from_slice(&self.version().to_le_bytes());
+        buffer.extend_from_slice(&marker_bytes);
+
+        Ok(buffer)
+    }
+
+    /// Deserializes from bytes, creating appropriate variant based on version.
+    fn from_bytes(data: &[u8]) -> Result<Self, bincode::Error> {
+        const VERSION_SIZE: usize = 2;
+
+        let version = u16::from_le_bytes(
+            data.get(..VERSION_SIZE)
+                .and_then(|bytes| bytes.try_into().ok())
+                .ok_or_else(|| bincode::Error::new(bincode::ErrorKind::SizeLimit))?,
+        );
+
+        let marker_data = &data[VERSION_SIZE..];
+
+        match version {
+            1 => Ok(Self::Current(BlockMarkerV1::from_bytes(marker_data)?)),
+            2 => Ok(Self::V2(BlockMarkerV2::from_bytes(marker_data)?)),
+            _ => Err(bincode::Error::new(bincode::ErrorKind::Custom(format!(
+                "Unsupported VersionedBlockMarker version: {version}"
+            )))),
+        }
+    }
+}
+
+impl Serialize for VersionedBlockMarker {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let bytes = self.to_bytes().map_err(serde::ser::Error::custom)?;
+        serializer.serialize_bytes(&bytes)
+    }
+}
+
+impl<'de> Deserialize<'de> for VersionedBlockMarker {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct VersionedBlockMarkerVisitor;
+
+        impl Visitor<'_> for VersionedBlockMarkerVisitor {
+            type Value = VersionedBlockMarker;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a serialized VersionedBlockMarker byte stream")
+            }
+
+            fn visit_bytes<E>(self, value: &[u8]) -> Result<VersionedBlockMarker, E>
+            where
+                E: de::Error,
+            {
+                VersionedBlockMarker::from_bytes(value).map_err(de::Error::custom)
+            }
+        }
+
+        deserializer.deserialize_bytes(VersionedBlockMarkerVisitor)
+    }
+}
+
+// ============================================================================
+// BlockMarkerV1 Implementation
+// ============================================================================
+
+impl BlockMarkerV1 {
+    /// Serializes to bytes with variant prefix.
+    fn to_bytes(&self) -> Result<Vec<u8>, bincode::Error> {
+        let Self::BlockFooter(footer) = self;
+        let footer_bytes = footer.to_bytes()?;
+
+        let mut buffer = Vec::with_capacity(1 + footer_bytes.len());
```

Comment:
nit: maybe we should put these constants describing length of a field somewhere near each other?